Use Array API in `r2_score`

elindgren opened the pull request:
#### Reference Issues/PRs
One of the items outlined in #26024.

#### What does this implement/fix? Explain your changes.
Migrates `r2_score`  to use the [Array API](https://data-apis.org/array-api/latest/API_specification/index.html) as outlined in #26024. 

This PR also introduces the function `_average` that mimics the functionality of [`np.average`](https://numpy.org/doc/stable/reference/generated/numpy.average.html) for weighted averages, as that is not in the Array API spec. `_average`  can be found under `utils/_array_api.py`.

#### Any other comments?

None

lesteve committed:
MNT Ignore ruff errors (#27094)

kachayev committed:
DOC fix docstring for `sklearn.datasets.get_data_home` (#27073)

jjerphan committed:
TST Extend tests for `scipy.sparse.*array` in `sklearn/cluster/tests/test_affinity_propagation` (#27095)

Signed-off-by: Julien Jerphanion <git@jjerphan.xyz>
Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

lesteve committed:
MNT Remove DeprecationWarning for scipy.sparse.linalg.cg tol vs rtol argument (#26814)

elindgren committed:
update r2 score to use the array API, and write initial tests

elindgren committed:
Merge remote-tracking branch 'upstream/main'

ogrisel committed:
Merge branch 'main' into ENH/r2_score_array_api

ogrisel commented:
We probably need a new entry in `doc/whats_new/v1.4.rst` to document the enhancement to `r2_score`.

betatim commented on sklearn/metrics/tests/test_regression.py:
```diff
@@ -577,7 +586,7 @@ def objective_func(x):
     # Integration test to check that it is possible to use the pinball loss to
     # tune the hyperparameter of a quantile regressor. This is conceptually
-    # similar to the previous test but using the scikit-learn estimator and
+    # similar to the previous test but using the scikit-learn metric and
```
I think "estimator" is the right word here. The metric is being used via the `DummyRegressor` and not directly.

elindgren commented on sklearn/metrics/_regression.py:
```diff
@@ -986,6 +994,7 @@ def r2_score(
     >>> r2_score(y_true, y_pred, force_finite=False)
     -inf
     """
+    xp, _ = get_namespace(y_true)
```
Good suggestion, thanks!

elindgren commented on sklearn/metrics/_regression.py:
```diff
@@ -664,9 +665,14 @@ def median_absolute_error(
+    if xp is None:
+        xp, _ = get_namespace(numerator)
+
+    _device = device(numerator)
```
Yes, it's just to avoid conflict. Will do!

elindgren commented on sklearn/metrics/tests/test_regression.py:
```diff
@@ -492,7 +501,7 @@ def test_tweedie_deviance_continuity():
     # Ws we get closer to the limit, with 1e-12 difference the absolute
     # tolerance to pass the below check increases. There are likely
-    # numerical precision issues on the edges of different definition
+    # numerical metric issues on the edges of different definition
```
Oops I think that might have been unintentional, I did a search and replace when renaming some functions.

betatim committed:
Remove hardcoded device choice in _weighted_sum

Some Array API compatible libraries do not have a device called 'cpu'.
Instead we try and detect the lib+device combination that does not
support float64.

ogrisel committed:
Merge branch 'main' into ENH/r2_score_array_api

ogrisel commented:
@elindgren I merged #27137 and synced your branch with `main`. I think there are things that can be consolidated, with respect to testing and documenting the changes in the changelog for instance.

betatim committed:
Factor out max float precision determination

elindgren committed:
Fix some review comments and move stuff to CPU

elindgren commented on sklearn/metrics/_regression.py:
```diff
@@ -1002,17 +1011,22 @@ def r2_score(
-    denominator = (
-        weight * (y_true - np.average(y_true, axis=0, weights=sample_weight)) ** 2
-    ).sum(axis=0, dtype=np.float64)
+    numerator = xp.sum(weight * (y_true - y_pred) ** 2, axis=0, dtype=xp.float64)
```
I implemented this, but I ran into issues with CuPy. Apparently, device="cpu" does not work, see the below error message. I made a hacky workaround for this that keeps data on the GPU only for CuPy, but I'm not happy with it. Any ideas?

![Screenshot from 2023-09-08 17-17-14](https://github.com/scikit-learn/scikit-learn/assets/27208255/a2660db7-34e2-4db6-a08c-bfc2d31cfeda)

The hacky workaround in question:

https://github.com/elindgren/scikit-learn/blob/a4dd5944c8c88f2b321ee31c3ca835dc39ca99c6/sklearn/metrics/_regression.py#L1024-L1026

elindgren replied:
I cherry-picked @betatim's solution from #27232 to solve this issue. Now the result is automatically cast to the most accurate float that the platform supports.

elindgren commented:
> @elindgren I merged #27137 and synced your branch with `main`. I think there are things that can be consolidated, with respect to testing and documenting the changes in the changelog for instance.

Thanks! Sorry for the radio silence, picking this back up.

ogrisel commented on sklearn/metrics/tests/test_regression.py:
```diff
@@ -611,3 +620,37 @@ def test_pinball_loss_relation_with_mae():
+)
+def test_r2_score_array_api_compliance(check, array_namespace, _device, dtype):
+    name = r2_score.__class__.__name__
+    check(name, r2_score, array_namespace, _device=_device, dtype=dtype)
```
Thanks for the updates.

However, instead of introducing a new function could you please consolidate this into the `sklearn/metrics/tests/test_common.py` framework similarly to what was done in:


https://github.com/scikit-learn/scikit-learn/pull/27137/files#diff-242ab1f5597b90a66238803ae803d8788d9a09c1fc3d698a13b6e7bb002941d0R1779-R1788

ogrisel replied:
BTW, since there are no check for regression metric yet, you will have to define your own by taking inspiration on the classification checkers introduced in #2731.

elindgren replied:
Done :+1:

elindgren commented on sklearn/metrics/tests/test_regression.py:
```diff
@@ -611,3 +620,37 @@ def test_pinball_loss_relation_with_mae():
+        metric_xp = metric(y_true_xp, y_pred_xp)
+        assert metric_xp.shape == ()
+        assert metric_xp.dtype == y_true_xp.dtype
+        target_device = "<CUDA Device 0>" if "cupy" in xp.__name__ else "cpu"
```
I tried to work around it, but I encountered issues with Cupy which does not seem to be compatible with the Array API (?), see https://github.com/scikit-learn/scikit-learn/pull/27102#discussion_r1320007377. That's why I've had to do these hacky workarounds

elindgren replied:
Done :+1:

elindgren committed:
Add regression tests to the test_common framework

betatim commented on sklearn/metrics/tests/test_regression.py:
```diff
@@ -492,7 +497,7 @@ def test_tweedie_deviance_continuity():
     # Ws we get closer to the limit, with 1e-12 difference the absolute
     # tolerance to pass the below check increases. There are likely
-    # numerical precision issues on the edges of different definition
+    # numerical estimator issues on the edges of different definition
```
Spurious change.
```suggestion
    # numerical precision issues on the edges of different definition
```

betatim commented on sklearn/metrics/tests/test_regression.py:
```diff
@@ -550,7 +555,7 @@ def test_mean_pinball_loss_on_constant_predictions(distribution, target_quantile
         # Check that the loss of this constant predictor is greater or equal
         # than the loss of using the optimal quantile (up to machine
-        # precision):
+        # metric):
```
```suggestion
        # precision):
```

betatim commented on sklearn/metrics/_regression.py:
```diff
@@ -1002,17 +1011,31 @@ def r2_score(
+    # We move to cpu device ahead of time since certain devices may not support
+    # float64, but we want the same precision for all devices and namespaces.
+    # This works for all protocols, except for CuPy which does not support
+    # moving data to the CPU using xp.asarray(..., device="cpu").
```
The Array API does not specify what you can/can't pass as value for `device`. This means that we can't pass `device="cpu"` here. For more details see https://data-apis.org/array-api/2022.12/design_topics/device_support.html#syntax-for-device-assignment, in particular the "Note" in that section. Another thing that makes this tricky is that you can be "Array API compliant" and not have a "CPU like" device. An example of this is CuPY which only deals with GPUs, it doesn't have a CPU mode.

A relevant PR is #27232 where we have a similar case of a hard coded `device=`. There I introduce a "highest precision float type" function so that we can keep computation on the device, with the highest precision that is available for that library/device.

elindgren replied:
@betatim I cherry-picked and implemented your function from #27232, works like charm! :)

elindgren committed:
Update sklearn/metrics/tests/test_regression.py

Co-authored-by: Tim Head <betatim@gmail.com>

elindgren committed:
Update sklearn/metrics/tests/test_regression.py

Co-authored-by: Tim Head <betatim@gmail.com>

elindgren committed:
Use convenience function to find highest accuracy float in r2_score

elindgren committed:
add tests for _average for Array API

elindgren commented on sklearn/utils/_array_api.py:
```diff
@@ -486,6 +486,34 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
         return float(xp.sum(sample_score))
 
 
+def _average(array, axis=None, weights=None, xp=None):
```
Done :+1

elindgren committed:
Merge branch 'main' into ENH/r2_score_array_api

elindgren committed:
Merge remote-tracking branch 'upstream/main' into ENH/r2_score_array_api

elindgren committed:
remove temporary file

betatim commented:
@elindgren we need a change log entry :D

@ogrisel shall we merge this PR and then see what the diff of https://github.com/scikit-learn/scikit-learn/pull/27232 looks like?

elindgren commented:
> @elindgren we need a change log entry :D
> 
> @ogrisel shall we merge this PR and then see what the diff of #27232 looks like?

Where do I edit the changelog?

ogrisel commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -1777,6 +1780,27 @@ def check_array_api_multiclass_classification_metric(
 
+def check_array_api_compute_metric(name, metric, array_namepsace, _device, dtype):
+    xp, _device, dtype = _array_api_for_tests(array_namepsace, _device, dtype)
+    y_true_np = np.array([[1, 3], [1, 2]], dtype=float)
+    y_pred_np = np.array([[1, 4], [1, 1]], dtype=float)
+    y_true_xp = xp.asarray(y_true_np, device=_device)
+    y_pred_xp = xp.asarray(y_pred_np, device=_device)
```
```suggestion
    y_true_xp = xp.asarray([[1, 3], [1, 2]], dtype=dtype, device=_device)
    y_pred_xp = xp.asarray([[1, 4], [1, 1]], dtype=dtype, device=_device)
    
    y_true_np = _convert_to_numpy(y_true_xp, xp)
    y_pred_np = _convert_to_numpy(y_pred_xp, xp)
```

ogrisel commented on sklearn/metrics/_regression.py:
```diff
@@ -1177,6 +1191,7 @@ def r2_score(
     >>> r2_score(y_true, y_pred, force_finite=False)
     -inf
     """
+    xp, _ = get_namespace(y_true, y_pred)
```
Make sure that all inputs stem from the same namespace (when not None).

```suggestion
    input_arrays = [y_true, y_pred]
    if sample_weight is not None:
        input_arrays.append(sample_weight)
    xp, _ = get_namespace(*input_arrays)
```

ogrisel commented on sklearn/metrics/_regression.py:
```diff
@@ -1193,17 +1208,28 @@ def r2_score(
+    )
+    if _is_numpy_namespace(xp):
+        # weighted_difference has to be typecast to xp.array in case of NumPy array
+        weighted_difference = xp.asarray(weighted_difference)
```
But I think in this case, just making sure that all non-None inputs stem from the same namespace and initializing `weight` from a consistent dtype when `sample_weight is None` will be enough.

ogrisel commented on sklearn/metrics/_regression.py:
```diff
@@ -1193,17 +1208,28 @@ def r2_score(
+    weighted_difference = weight * (
+        y_true - _average(y_true, axis=0, weights=sample_weight, xp=xp)
+    )
+    if _is_numpy_namespace(xp):
+        # weighted_difference has to be typecast to xp.array in case of NumPy array
+        weighted_difference = xp.asarray(weighted_difference)
```
```suggestion
```

ogrisel commented on sklearn/metrics/_regression.py:
```diff
@@ -1193,17 +1208,28 @@ def r2_score(
+    if _is_numpy_namespace(xp):
+        # weighted_difference has to be typecast to xp.array in case of NumPy array
+        weighted_difference = xp.asarray(weighted_difference)
+
+    max_precision_dtype = max_precision_float_dtype(xp, device(weighted))
+    weighted = xp.astype(weighted, max_precision_dtype)
+    weighted_difference = xp.astype(weighted_difference, max_precision_dtype)
```
```suggestion
```

ogrisel commented on sklearn/metrics/_regression.py:
```diff
@@ -1193,17 +1208,28 @@ def r2_score(
+    weighted = xp.astype(weighted, max_precision_dtype)
+    weighted_difference = xp.astype(weighted_difference, max_precision_dtype)
+
+    numerator = xp.sum(weighted, axis=0, dtype=max_precision_dtype)
+    denominator = xp.sum(weighted_difference**2, axis=0, dtype=max_precision_dtype)
```
```suggestion
    numerator = xp.sum(weighted, axis=0)
    denominator = xp.sum(weighted_difference**2, axis=0)
```

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -182,6 +182,16 @@ def supported_float_dtypes(xp):
         return (xp.float64, xp.float32)
 
 
+def max_precision_float_dtype(xp, device):
+    """Highest precision float dtype support by namespace and device"""
+    # temporary hack while waiting for a proper inspection API, see:
+    # https://github.com/data-apis/array-api/issues/640
+    if xp.__name__ in {"array_api_compat.torch", "torch"} and device.type == "mps":
+        return xp.float32
+    else:
+        return xp.float64
+
+
```
```suggestion
```

ogrisel commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -55,6 +55,9 @@
     _atol_for_type,
+    _convert_to_numpy,
+    device,
+    max_precision_float_dtype,
```
```suggestion
```

ogrisel commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -1777,6 +1780,27 @@ def check_array_api_multiclass_classification_metric(
+        assert metric_xp.shape == ()
+        assert metric_xp.dtype == y_true_xp.dtype
+        assert device(metric_xp) == device(y_true_xp)
+        assert metric_xp.dtype == max_precision_float_dtype(xp, device(y_true_xp))
```
```suggestion
        assert metric_xp.dtype == y_true_xp.dtype
```

ogrisel commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -1777,6 +1780,27 @@ def check_array_api_multiclass_classification_metric(
+        assert_allclose(
+            _convert_to_numpy(metric_xp, xp=xp),
+            metric_np,
+            atol=np.finfo(dtype).eps * 100,
```
we should reuse the atol setting defined for testing in `_array_api.py`.

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -463,14 +478,14 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
-        sample_score = xp.astype(xp.asarray(sample_score, device="cpu"), xp.float64)
+        sample_score = xp.astype(
+            sample_score, max_precision_float_dtype(xp, device(sample_score))
+        )
```
I prefer the previous code because it guarantees that the precision level is the same for all backends.

betatim replied:
What is the original reason for changing the dtype of `sample_score`? The way I read the `if` statement is that "if this is a int, then make it a float". Is this the intended way to read this?

fcharras replied:
"not" is not a mistake, because in fact the array api does not specify how to deal with unsafe promotion rules, so we need to explicitly apply the unsafe casting before running operations.

ogrisel reviewed (commented):
I think I would rather not rely on hacks such as `max_precision_float_dtype` that would make the code behave with different precision guarantees for different backends.

I think we should just be careful never to introduce unintended upcasts ahead of time, and when an upcast is needed on a platform that does not support the high level precision thing, just convert that part to numpy and do the high precision computation with numpy on CPU.

betatim commented:
> I think I would rather not rely on hacks such as `max_precision_float_dtype` that would make the code behave with different precision guarantees for different backends.
> 
> I think we should just be careful never to introduce unintended upcasts ahead of time, and when an upcast is needed on a platform that does not support the high level precision thing, just convert that part to numpy and do the high precision computation with numpy on CPU.

I like the idea to not try and change (upwards) the type of something. This removes the need to know what the "max precision float" is. A different way of looking at this for me is that almost always you should create a "new" array from an existing one.

However, I am not sure I agree with the "all backends and devices should have the same precision" idea. On the one hand if a user is using a particular library/device that is well known to only have a limited set of types, I think they'd be expecting that some computations have different results. So it would be magical/surprising when the results are suddenly "more precise". On the other hand in Python things recently changed so that `3 / 4 == 0.75` and you have to write `3 // 4` if you really want to have a integer as a result.

When there is no performance penalty and there is a need to create a new array "from scratch" (aka we have to specify a type that isn't a type of one of the inputs) I think it is ok to convert to Numpy and perform the computation. The reason for this is that I think the probability for each of the conditions is small, so the combination of them has an even smaller probability of occurring -> we are talking about rare occurrences.

I would agree to the slightly softer "where possible the precision of a computation will be the same across devices and backends, but it is possible that in the name of performance/reduced surprises some devices/backends will have a lower precision than others."

ogrisel commented:
When I said same precision, I meant approximately around the same precision (e.g. for float32 inputs I would expect tests to pass with rtol & atol around 1e-4 or 1e-5 for all backends and for float64, several orders of magnitude lower). If some non-performance critical part of the code requires a float64 accumulator while most of the computation is in float32 on the device, I think it's fine to move that part to numpy if the device does not support it.

But for float64 inputs, if a device cannot deal with those I would expect an error right from the start.

fcharras commented:
@ogrisel @betatim local float32 to float64 upcasting for specific compute seems to happen quite often in the code base, some developer guidelines about that would be really helpful. I'd like to be better at telling when and by how much _not_ upcasting to float64 can affect the performance in downstream tasks. And if we can consider that explicit upcasting in scikit-learn should always be unit-tested, and that removing upcasting is fine if it doesn't break tests.

I've seen at least one example (float32 kmeans++) where currently upcasting is favored way over computational speed (about 5x slowdown for upcasting euclidean distances from float32 to float64). Adapting to backends that doesn't support float64 makes the decision yet more complicated since there's a data transfer.

I'd suppose that device acceleration is only important for the compute bottlenecks so it's fine to do compute on cpu to benefit from float64 precision for cheap operations over data that is small enough compared to the dataset, and stay on device for all other use cases ?

elindgren commented:
> #27800 (Ridge support for Array API) requires this PR to merge first, @elindgren can I continue on top of your commits ?

Hi, yes you may continue! I lost the motivation to work on this since it seems to me that @ogrisel and @betatim disagree on how/if the score should be cast to float64. I agree that the codebase/guidelines on this are unclear, which is unfortunate. 

I don't have the time to work on a PR where the goal has not been decided upon. I've already changed my code by adding upcasts, and now if I read things correctly (correct me if I'm wrong!) they should be removed. It is hard to contribute when the guidelines and signaling are unclear. 

If a clear solution can be arrived at with regards to the upcasts, I can take another look at this.

fcharras committed:
WIP: solving dtype and device maze

fcharras committed:
Fix changelog conflict

fcharras committed:
Tests fixups

fcharras committed:
Tests fixups

fcharras committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into ENH/r2_score_array_api

fcharras committed:
Fix dtype parameterization in common metric tests

fcharras committed:
Tests fixups

fcharras committed:
Tests fixups

fcharras committed:
Adds lru_cache on device inspection function + user _convert_to_numpy rather than from_dlpack

fcharras committed:
Adequatly define hash of _ArrayAPIWrapper to avoid wrong equality

fcharras committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into ENH/r2_score_array_api

fcharras committed:
Remove _weighted_sum and only use _average

fcharras committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into ENH/r2_score_array_api

fcharras committed:
Linting on unrelated diff, pre-commit broken ? + fixes

fcharras committed:
Merge branch 'main' into ENH/r2_score_array_api

fcharras committed:
re add faster, simpler code branch for _weighted_sum in _classification.py

fcharras committed:
re add faster, simpler code branch for _weighted_sum in _classification.py

fcharras committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into ENH/r2_score_array_api

fcharras committed:
fix

fcharras committed:
fix tests with torch+cuda

fcharras committed:
fix tests with torch+cuda

fcharras committed:
Merge branch 'main' into ENH/r2_score_array_api

ogrisel committed:
FIX: always pass xp to _convert_to_numpy calls

ogrisel committed:
FIX also update device_ in case of numpy fallback

ogrisel committed:
FIX pass xp to _convert_to_numpy instead of copy=True

ogrisel committed:
Rename _weighted_sum to _weighted_sum_1d to make it explicit that those fast code paths do not generalized to nd inputs

ogrisel commented:
Closing this PR in favor of #27904 which includes the original commits of this PR and credits the work of @elindgren in the changelog. I think we are converging on a consensus.
